internal/contributebot/design/merge.md

```diff
@@ -0,0 +1,98 @@
+# Contribute Bot Auto-Merge Design
+
+Author: [@zombiezen][]<br>
+Date: 2018-12-17<br>
+Part of [#687][]
+
+[@zombiezen]: https://github.com/zombiezen
+[#687]: https://github.com/google/go-cloud/issues/687
+
+## Objective
+
+As a maintainer of Go Cloud, it is desirable to have pull requests merged in as
+quickly as they can be merged after receiving approvals, assuming there are no
+test breakages or merge conflicts. Currently, merging requires a manual process:
+
+```
+while branch is not up-to-date with master {
+  press update branch button
+  wait for Travis run to finish
+}
+click squash-and-merge
+```
+
+This loop is especially expensive the longer the waiting is, since it increases
+the likelihood that the branch is not up-to-date and thus causing the cycle to
+increase iterations (more waiting!).
+
+The objective of this design is to reduce the amount of maintainer [toil][] in
+this process.
+
+[toil]: https://landing.google.com/sre/sre-book/chapters/eliminating-toil/
+
+## Overview
+
+To reduce maintainer toil, this document proposes a new set of interactions for
+Contribute Bot. If a user writes a comment on a pull request with the text
+"/merge", Contribute Bot will:
+
+1.  Verify that the commenter has `write` or `admin` access to the repository.
+    If they do not, do not proceed and leave an error comment on the PR.
+2.  Verify that the pull request has the necessary approvals. If it does not, do
+    not proceed and leave an error comment on the PR.
+3.  Verify that there are no other commits on the pull request branch since the
+    last approved commit (or a manually entered commit hash entered by the
+    commenter) other than those created by Contribute Bot. If such commits
+    exists, then do not proceed and leave a comment on the PR mentioning that
+    "/merge COMMITHASH" will bypass this. While this is stricter than what the
+    branch protections might allow, it avoids accidentally merging in unreviewed
+    changes.
+3.  Leave a comment on the PR to the effect that Contribute Bot has received the
+    request and that it will report back if it is unable to merge the PR.
+4.  If the pull request is not up to date with the changes on the base branch,
+    then create a merge commit on the pull request branch. If the merge commit
+    cannot be created due to merge conflicts, then stop trying to merge the PR
+    and make an error comment on the PR.
+5.  Wait for any necessary check runs to report completion. If any report
+    failure, then stop trying to merge the PR and make an error comment on the
+    PR.
+6.  Merge the pull request using the project's allowed merge method (`squash`
+    for Go Cloud and Wire), using the body of the pull request (the first
+    comment) as the commit message in the case of `merge` or `squash`.
+
+Multiple pull requests may be requested to be merged at the same time. In this
+case, Contribute Bot will hold off on proceeding through Steps 4-6 until the
```

Comment:
Just to clarify: if ContributeBot gets 3 `/merge` requests at the same time, and all 3 PRs require merging, will it:

- Attempt to merge one of them, and block the other 2 until that completes?
- Or, merge all 3, and end up submitting whichever one's tests finish first (and then merge the other 2 again, rinse and repeat)?

The former seems more efficient but might be harder to implement. Related: how will ContributeBot manage state across restarts?

Reply:
The first interpretation is the one I intend. I'm not sure what verbage to add to make that more clear.

AFAICT it's about the same effort to do either. My plan is to keep some additional state in a SQL database (seems necessary regardless), but I was holding off on detailing that until the overall flow was approved.

Reply:
The other state I had in mind was detecting which merge commits came from the bot. I suppose we could remove that state if we had the bot GPG sign commits, because it could verify the same identity later.

The worker could then keep the state of active merge requests in memory, then do a scan of all open pull requests on restarts with the "/merge" comments (or potentially a label). Doable, but could make restarts expensive (failure mode could be blowing out GitHub API quota).